cmd/kwil-admin/cmds/snapshot/create.go

```diff
@@ -233,8 +233,28 @@ func pgDump(ctx context.Context, dbName, dbUser, dbPass, dbHost, dbPort string,
 				Name:   fmt.Sprintf("validator-%d", validatorCount),
 			})
 			validatorCount++
+		} else if inFunctionBlock {
+			if strings.HasPrefix(line, "$$;") {
```

Comment:
I'm trying to find documentation about this:
- can there be other functions that end in something other than `$$`;?  like functions that happen to be in the dump that we didn't define for procedures.  We might scan all the way to the end, or just too far (until the next function end).
- Why prefix vs suffix vs equals (ignoring leading/trailing space)?
- we often define functions ending with `$$ LANGUAGE plpgsql;` but I guess that gets simplified.

Reply:
https://www.postgresql.org/docs/current/plpgsql-structure.html

Reply:
Fortunately postgresql will hang/error if trying to do this:

```
CREATE FUNCTION test_func() RETURNS text AS $$
BEGIN
    RETURN 'This string contains $$ inside the body';
END;
$$ LANGUAGE plpgsql;
```

The only way to do that is to change the delimiter (`$$`) to something else:

```
CREATE FUNCTION test_func() RETURNS text AS $delim$
BEGIN
    RETURN 'This string contains $$ inside the body';
END;
$delim$ LANGUAGE plpgsql;
```

When we are crafting the `CREATE FUNCTION` in kwil, we use `$$`, so the function body cannot contain "$$" anywhere.

---

The remaining question I have is will pg_dump include any other system created functions that might use a different delimiter.  (It creates a bunch of selects and sets to get things right like locate, so who knows if it's also going to create a function.

I think this will work for now, but we should document this pretty clearly.

Reply:
```
// create_user creates a user in the database.
// It is assigned a unique uuid.
// MUTABLE
procedure create_user($name text, $age int) public {
    // we will generate a uuid from the txid
    notice('stuff $$ adsf');
    INSERT INTO users (id, name, age, address)
    VALUES (uuid_generate_v5('985b93a4-2045-44d6-bde4-442a4e498bc6'::uuid, @txid), $name, $age, @caller);
}
```

on deploy:

`failed to deploy database, code = 4294967295, log = "ERROR: syntax error at or near \"adsf\" (SQLSTATE 42601)"`